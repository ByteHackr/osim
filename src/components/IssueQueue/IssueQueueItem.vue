<script setup lang="ts">
import { computed } from 'vue';

import { useToggle } from '@vueuse/core';

import type { FilteredIssue } from './IssueQueue.vue';

const { issue } = defineProps<{
  issue: FilteredIssue;
}>();

// Temporarily hiding 'Source' column to avoid displaying incorrect information.
// TODO: unhide it once final issue sources are defined. [OSIDB-2424]
//       and update the CSS for the column width in IssueQueue

const nonIdFields: Exclude<keyof FilteredIssue, 'id'>[] =
  ['impact', 'formattedDate', 'title', 'workflowState', 'owner'];

const hasBadges = computed(() => issue.embargoed || !!issue.labels?.length);
const truncatedLabelCount = 4;
const [showLabels, toggleShowLabels] = useToggle();
const sortedLabels = computed(() => issue.labels?.toSorted((a, b) => {
  if (
    (a.state === 'REQ' && b.state !== 'REQ')
    || (a.contributor && !b.contributor)
  ) {
    return -1;
  }
  if (
    (a.state !== 'REQ' && b.state === 'REQ')
    || (!a.contributor && b.contributor)
  ) {
    return 1;
  }

  return a.label.localeCompare(b.label);
}) ?? []);
const truncatedLabels = computed(
  () => showLabels.value ? sortedLabels.value : sortedLabels.value?.slice(0, truncatedLabelCount),
);
</script>

<template>
  <tr
    class="osim-issue-queue-item"
    :class="$attrs.class"
  >
    <td
      class="osim-issue-title"
      :class="{ 'pb-0': hasBadges }"
    >
      <RouterLink :to="{ name: 'flaw-details', params: { id: issue.id } }">
        {{ issue.id }}
      </RouterLink>
    </td>
    <td
      v-for="field in nonIdFields"
      :key="field"
      :class="{ 'pb-0': hasBadges }"
    >
      {{ issue[field] }}
    </td>
    <!--<td>{{ issue.assigned }}</td>-->
  </tr>
  <tr
    v-if="hasBadges"
    class="osim-badge-lane"
    :class="$attrs.class"
  >
    <td colspan="100%">
      <div class="gap-1 d-flex">
        <span
          v-if="issue.embargoed"
          class="badge rounded-pill text-bg-danger border border-primary"
        >Embargoed</span>
        <span
          v-for="label in truncatedLabels"
          :key="label.label"
          class="badge rounded-pill text-capitalize border"
          :class="{
            'text-bg-warning fw-bold border-warning': label.state == 'REQ',
            'text-bg-light-info border-info': label.state != 'REQ' && label.relevant,
            'text-decoration-line-through text-bg-gray border-secondary': !label.relevant,
          }"
          :title="label.state == 'REQ' ? 'Requested' : ''"
        >{{ label.label }}</span>
        <i
          v-if="(issue.labels?.length ?? 0) > truncatedLabelCount"
          class="bi pe-1 cursor-pointer osim-show-all-labels"
          :class="[showLabels ? 'bi-caret-left-fill' : 'bi-caret-right-fill']"
          title="Show all labels"
          @click="toggleShowLabels()"
        />
      </div>
    </td>
  </tr>
</template>

<style lang="scss" scoped>
@import '@/scss/bootstrap-overrides';

td.osim-issue-title {
  max-width: 10rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

tr td {
  border: 0;
}

tr.osim-shaded td {
  background-color: $light-gray;
}

tr:hover {
  td {
    background-color: #ffe3d9;
  }

  &.osim-issue-queue-item + tr.osim-badge-lane td {
    background-color: #ffe3d9;
  }
}

tr.osim-issue-queue-item:has(+ tr.osim-badge-lane:hover) td {
  background-color: #ffe3d9;
}

.osim-badge-lane {
  .osim-show-all-labels {
    cursor: pointer;
  }
}
</style>
