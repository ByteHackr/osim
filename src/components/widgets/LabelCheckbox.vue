<script setup lang="ts">
const props = defineProps<{
  modelValue: boolean,
  label: string,
}>();

</script>

<template>
  <div class="osim-input border-start ps-3">
  <label class="form-check mb-3">
    <input
        type="checkbox"
        :checked="modelValue"
        class="form-check-input"
        v-bind="$attrs"
        :value="modelValue"
        @change="$emit('update:modelValue', (($event as InputEvent).target as HTMLInputElement).checked)"
    />
    <span class="form-check-label">
      {{ label }}
    </span>
  </label>
  </div>
</template>

<style scoped>

.osim-input {
  display: block;
}

</style>
