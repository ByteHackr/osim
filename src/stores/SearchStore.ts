import { computed } from 'vue';

import { defineStore } from 'pinia';
import { useLocalStorage } from '@vueuse/core';

export type SearchSchema = {
  queryFilter: string;
  searchFilters: Record<string, string>;
};

const defaultValues: SearchSchema = { searchFilters: {}, queryFilter: '' };

const _searchStoreKey = 'SearchStore';
const search = useLocalStorage(_searchStoreKey, defaultValues);
function saveFilter(filters: Record<string, string>, query: string = '') {
  search.value.searchFilters = filters;
  search.value.queryFilter = query;
}

function resetFilter() {
  search.value.searchFilters = {};
  search.value.queryFilter = '';
}
export const useSearchStore = defineStore('SearchStore', () => {
  const searchFilters = computed(() => search.value.searchFilters || {});
  const queryFilter = computed(() => search.value.queryFilter || '');

  return {
    searchFilters,
    queryFilter,
    saveFilter,
    resetFilter,
  };
});
