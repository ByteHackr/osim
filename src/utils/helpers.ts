import { toRaw, isRef, isReactive, isProxy, ref, toRef, watch } from 'vue';
import * as R from 'ramda';

export function vRef(prop: Record<string, any>, property: string, defaultValue: any) {
  const reffedProp = toRef(prop, property);
  watch(reffedProp, (value) => flexRef.value = value);
  const flexRef = reffedProp.value === undefined ? ref(defaultValue) : reffedProp;
  return flexRef;
}

export function deepToRaw<T extends Record<string, any>>(sourceObj: T): T {
  const objectIterator = (input: any): any => {
    if (Array.isArray(input)) {
      return input.map((item) => objectIterator(item));
    }
    if (isRef(input) || isReactive(input) || isProxy(input)) {
      return objectIterator(toRaw(input));
    }
    if (input && typeof input === 'object') {
      return Object.keys(input).reduce((acc, key) => {
        acc[key as keyof typeof acc] = objectIterator(input[key]);
        return acc;
      }, {} as T);
    }
    return input;
  };

  return objectIterator(sourceObj);
}

export function deepCopyFromRaw<T extends Record<string, any>>(sourceObj: T): T {
  return JSON.parse(JSON.stringify(deepToRaw(sourceObj)));
}

// eslint-disable-next-line max-len
// https://stackoverflow.com/questions/14446511/most-efficient-method-to-groupby-on-an-array-of-objects
export const groupBy = <T>(
  array: T[],
  predicate: (value: T, index: number, array: T[]) => string,
) =>
    array.reduce(
      (acc, value, index, array) => {
        (acc[predicate(value, index, array)] ||= []).push(value);
        return acc;
      },
    {} as { [key: string]: T[] },
    );

export const assignKeyValue = (object: Record<string, any>, key: string, value: any = null) => {
  object[key] = value;
  return object;
};

export const objectMap = (object: Record<string, any>, mapFn: (key: string, value: any) => any) =>
  Object.keys(object).reduce(
    (acc, key) => assignKeyValue(acc, key, mapFn(key, object[key])),
    {} as Record<string, any>,
  );

export const sortedByGroup = <T extends Record<string, any>>(array: T[], key: string) =>
  groupBy(
    array
      .filter((item: T) => item[key])
      .sort((itemA: T, itemB: T) => itemA[key].localeCompare(itemB[key])),
    (item: T) => item[key],
  );

type DeepMappable = any[] | Record<string, any>

const isNonEmptyArray = (value: any) => R.is(Array, value) && value.length > 0;
const isNonArrayObject = (value: any) => R.is(Object, value) && !R.is(Array, value);
const isDeepMappable = (value: DeepMappable) => isNonEmptyArray(value) || isNonArrayObject(value);

export const deepMap = (transform: (arg: any) => any, object: DeepMappable): any =>
  R.map(
    (val: any) => isDeepMappable(val)
      ? deepMap(transform, val)
      : transform(val)
    , object
  );

export const cveRegex = /^\s*(CVE-\d{4}-\d{4,7})\s*$/;
