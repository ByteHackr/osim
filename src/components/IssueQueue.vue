<script setup lang="ts">
import { computed, onMounted, reactive, ref, onUnmounted, watch } from 'vue';
import { DateTime } from 'luxon';
import IssueQueueItem from '@/components/IssueQueueItem.vue';
import LabelCheckbox from './widgets/LabelCheckbox.vue';
import { useUserStore } from '@/stores/UserStore';
import { FlawClassificationStateEnum } from '../generated-client';
const userStore = useUserStore();

const emit = defineEmits(['flaws:fetch', 'flaws:load-more']);

type FilteredIssue = {
  issue: any;
  selected: boolean;
};

type ColumnField = 'id' | 'impact' | 'source' | 'created_dt' | 'title' | 'state' | 'owner';

const props = defineProps<{
  issues: any[];
  isLoading: boolean;
  isFinalPageFetched: boolean;
}>();

const issues = computed<any[]>(() => props.issues.map(relevantFields));
const selectedSortField = ref<ColumnField>('created_dt');
const isSortedByAscending = ref(false);
const isMyIssuesSelected = ref(false);
const isOpenIssuesSelected = ref(false);
const tableContainerEl = ref<HTMLElement | null>(null);

const filteredStates = computed(() => {
  const allStates = Object.values(FlawClassificationStateEnum);
  return allStates
    .filter(
      (state) =>
        state !== FlawClassificationStateEnum.Done &&
        state !== FlawClassificationStateEnum.Rejected,
    )
    .join(',');
});

const params = computed(() => {
  const paramsObj: Record<string, any> = {};

  if (isMyIssuesSelected.value) {
    paramsObj.owner = userStore.userName;
  }

  if (isOpenIssuesSelected.value) {
    paramsObj.workflow_state = filteredStates.value;
  }

  const sortOrderPrefix = isSortedByAscending.value ? '' : '-';
  paramsObj.order = {
    [selectedSortField.value]: `${sortOrderPrefix}${selectedSortField.value}`,
    id: `${sortOrderPrefix}cve_id,${sortOrderPrefix}uuid`,
    state: `${sortOrderPrefix}workflow_state`,
  }[selectedSortField.value];
  return paramsObj;
});

const columnsFieldsMap: Record<string, ColumnField> = {
  ID: 'id',
  Impact: 'impact',
  Source: 'source',
  Created: 'created_dt',
  Title: 'title',
  State: 'state',
  Owner: 'owner',
};

const relevantIssues = computed<FilteredIssue[]>(() => {
  return issues.value
    .map((issue) => reactive({ issue, selected: false }));
});

function updateSelectAll(selectedAll: boolean) {
  for (const filteredIssue of relevantIssues.value) {
    filteredIssue.selected = selectedAll;
  }
}

function selectSortField(field: ColumnField) {
  isSortedByAscending.value =
    selectedSortField.value === field ? !isSortedByAscending.value : false;
  selectedSortField.value = field;
}

const isSelectAllIndeterminate = computed(() => {
  if (relevantIssues.value.length === 0) {
    return false;
  }
  return relevantIssues.value.some((it) => it.selected !== relevantIssues.value[0].selected);
});

const isSelectAllChecked = computed(() => {
  return relevantIssues.value.every((it) => it.selected);
});

function relevantFields(issue: any) {
  return {
    id: issue.cve_id || issue.uuid,
    impact: issue.impact,
    source: issue.source,
    created_dt: issue.created_dt,
    title: issue.title,
    workflowState: issue.classification.state,
    unembargo_dt: issue.unembargo_dt,
    embargoed: issue.embargoed,
    owner: issue.owner,
    formattedDate: DateTime.fromISO(issue.created_dt).toFormat('yyyy-MM-dd'),
  };
}

onMounted(() => {
  tableContainerEl.value?.addEventListener('scroll', handleScroll); // AutoScroll Option
  emit('flaws:fetch', params);
});

onUnmounted(() => {
  tableContainerEl.value?.removeEventListener('scroll', handleScroll);
});

// AutoScroll Method. Maybe have this as a user configurable option in the future?
function handleScroll() {
  if (props.isLoading || !tableContainerEl.value) return; // Do not load more if already loading

  const totalHeight = tableContainerEl.value.scrollHeight;
  const scrollPosition = tableContainerEl.value.scrollTop + tableContainerEl.value.clientHeight;
  // Trigger loading more content when the user has scrolled to 99% of the container's height, 
  if (scrollPosition >= totalHeight * 0.99) { 
    emitLoadMore();
  }
}

function emitLoadMore() {
  emit('flaws:load-more', params);
}

watch(params, () => {
  emit('flaws:fetch', params);
});
</script>

<template>
  <div class="osim-content container osim-issue-queue">
    <div class="osim-incident-filter">
      <LabelCheckbox v-model="isMyIssuesSelected" label="My Issues" class="d-inline-block" />
      <LabelCheckbox v-model="isOpenIssuesSelected" label="Open Issues" class="d-inline-block" />

      <span
        v-if="isLoading"
        class="spinner-border spinner-border-sm d-inline-block ms-3"
        role="status"
      >
        <span class="visually-hidden">Loading...</span>
      </span>
      <span v-if="isLoading"> Loading Flaws&hellip; </span>
    </div>
    <div ref="tableContainerEl" class="osim-incident-list">
      <table class="table align-middle" :class="{ 'osim-table-loading': isLoading }">
        <thead class="sticky-top">
          <!-- <thead class=""> -->
          <tr>
            <th>
              <input
                type="checkbox"
                class="form-check-input"
                :indeterminate="isSelectAllIndeterminate"
                :checked="isSelectAllChecked"
                aria-label="Select All Issues in Table"
                @change="updateSelectAll(($event.target as HTMLInputElement).checked)"
              />
            </th>
            <th
              v-for="(field, columnName) in columnsFieldsMap"
              :key="columnName"
              :class="{ 'osim-issue-queue-table-created': columnName === 'Created' }"
              @click="selectSortField(field)"
            >
              {{ columnName }}
              <i
                :class="{
                  'opacity-0': selectedSortField !== field,
                  'bi-caret-up-fill': isSortedByAscending,
                  'bi-caret-down-fill': !isSortedByAscending,
                }"
                class="bi"
              >
              </i>
            </th>
          </tr>
        </thead>
        <tbody class="table-group-divider">
          <template v-for="(relevantIssue, index) of relevantIssues" :key="relevantIssue.id">
            <IssueQueueItem
              v-model:selected="relevantIssue.selected"
              :issue="relevantIssue.issue"
              :class="{
                'osim-shaded': index % 2 === 0,
              }"
            />
          </template>
        </tbody>
      </table>
      <button
        v-if="relevantIssues.length !== 0 && !isFinalPageFetched"
        class="btn btn-primary"
        type="button"
        :disabled="isLoading"
        @click="emitLoadMore"
      >
        <span
          v-if="isLoading"
          class="spinner-border spinner-border-sm d-inline-block"
          role="status"
        >
          <span class="visually-hidden">Loading...</span>
        </span>
        <span v-if="isLoading"> Loading Flaws&hellip; </span>
        <span v-else> Load More Flaws </span>
      </button>
      <p v-if="relevantIssues.length === 0">No results.</p>
      <span v-if="isFinalPageFetched" role="status">Nothing else to load.</span>
    </div>
  </div>
</template>

<style scoped lang="scss">
@import '@/scss/redhat-brand-colors';

.osim-issue-queue {
  font-family: 'Red Hat Mono', monospace;

  // tbody

  div.osim-incident-list {
    display: block;
    max-height: calc(100vh - 164px);
    overflow-y: scroll;

    &:hover::-webkit-scrollbar-thumb {
      background-color: $redhat-red-50;
    }
  }

  .osim-incident-filter {
    margin-left: -0.5rem;
  }

  &:hover i {
    color: $redhat-red-50;
  }

  table {
    &.osim-table-loading {
      opacity: 0.5;
    }

    th {
      cursor: pointer;
      user-select: none;
    }

    tr td,
    tr th {
      &:not(:first-child) {
        min-width: 12ch;
      }
    }

    tbody:hover::-webkit-scrollbar-thumb {
      background-color: $redhat-red-50;
    }
  }
}
</style>
