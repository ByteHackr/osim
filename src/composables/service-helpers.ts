import { getDisplayedOsidbError } from '@/services/OsidbAuthService';
import { useToastStore } from '@/stores/ToastStore';
import { AxiosError, type AxiosResponse } from 'axios';

export function createCatchHandler(title: string = 'Error', shouldThrow: boolean = true) {
  return (error: AxiosError) => {
    const displayedError = getDisplayedOsidbError(error);
    const { addToast } = useToastStore();
    addToast({
      title,
      body: `${displayedError}: ${error.response?.data}`,
      css: 'warning',
    });
    console.error('‚ùå ', error);
    if (shouldThrow) throw error;
    return error;
  };
}

export function createSuccessHandler({ title, body }: Record<string, string>, onSucceed: Function = () => {}) {
  return (response: AxiosResponse) => {
    const { addToast } = useToastStore();
    onSucceed(response);
    addToast({
      title,
      body,
      css: 'success',
    });
  };
}
