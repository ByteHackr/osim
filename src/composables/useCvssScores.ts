import { putFlawCvssScores, postFlawCvssScores } from '@/services/FlawService';
import { type Flaw } from '@/generated-client';
import { createSuccessHandler, createCatchHandler } from './service-helpers';
import { computed, ref, watch, type Ref } from 'vue';

const onSucceed = createSuccessHandler({ title: 'Success!', body: 'Saved CVSS Scores' });
const onError = createCatchHandler('Error updating Flaw CVSS data');

export function useCvssScores(flaw: Ref<Flaw>) {
  const wasCvssModified = ref(false);
  const redHatCvssData = flaw.value.cvss_scores.find((assessment) => assessment.issuer === 'RH');
  const flawRhCvss = ref(redHatCvssData || flaw.value.cvss_scores[0]);
  
  watch(flawRhCvss, () => wasCvssModified.value = true, {deep: true});

  const nistCvssScore = computed(() => flaw.value.cvss_scores.find((score) => score.issuer === 'NIST'));
  const flawNvdCvssScore = computed(() => nistCvssScore.value?.score || flaw.value.nvd_cvss3);

  async function saveCvssScores() {
    if (flawRhCvss.value.created_dt) {
      return putFlawCvssScores(
        flaw.value.uuid,
        flawRhCvss.value.uuid || '',
        flawRhCvss.value as unknown
      )
        .then(onSucceed)
        .catch(onError);
    } else {
      const requestBody = {
        // "score":  is recalculated based on the vector by OSIDB and does not need to be included
        comment: flawRhCvss.value.comment,
        cvss_version: 'V3',
        issuer: 'RH',
        vector: flawRhCvss.value.vector,
        embargoed: flaw.value.embargoed,
      };
      return postFlawCvssScores(flaw.value.uuid, requestBody as unknown)
        .then(onSucceed)
        .catch(onError);
    }
  }

  return {
    wasCvssModified,
    flawNvdCvssScore,
    flawRhCvss,
    saveCvssScores,
  };
}

