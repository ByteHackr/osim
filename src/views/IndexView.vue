<script setup lang="ts">
import { computed, onDeactivated, ref, watch, type Ref } from 'vue';
import IssueQueue from '../components/IssueQueue.vue';
import { useFlawsFetching } from '../composables/useFlawsFetching';
import { useSearchStore } from '@/stores/SearchStore';

const searchStore = useSearchStore();

const { issues, isLoading, isFinalPageFetched, total, loadFlaws, loadMoreFlaws } = useFlawsFetching();

const tableFilters = ref<Record<string, string>>({});

const showFilter = computed(() =>
  Object.keys(searchStore.searchFilters).length > 0
);

const isDefaultFilterSelected = ref<boolean>(showFilter.value);

const params = computed(() => {
  const paramsObj = {
    ...tableFilters.value,
    ...isDefaultFilterSelected.value && searchStore.searchFilters
  };

  return paramsObj;
});

watch(() => params, () => {
  loadFlaws(params);
}, {
  deep: true
});

function fetchMoreFlaws() {
  loadMoreFlaws(params);
}

function setTableFilters(newFilters: Ref<Record<string, string>>) {
  tableFilters.value = {
    ...newFilters.value
  };
}

onDeactivated(() => {
  loadFlaws(params);
});
</script>

<template>
  <main class="mt-3">
    <IssueQueue
      v-model:isDefaultFilterSelected="isDefaultFilterSelected"
      :issues="issues"
      :isLoading="isLoading"
      :total="total"
      :isFinalPageFetched="isFinalPageFetched"
      :showFilter="showFilter"
      :defaultFilters="searchStore.searchFilters"
      @flaws:fetch="setTableFilters"
      @flaws:load-more="fetchMoreFlaws"
    />
  </main>
</template>
