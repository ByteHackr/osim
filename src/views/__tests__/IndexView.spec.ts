import { ref } from 'vue';

import { flushPromises, type VueWrapper } from '@vue/test-utils';

import IssueQueue from '@/components/IssueQueue.vue';
import LabelCheckbox from '@/components/widgets/LabelCheckbox.vue';

import { useFlawsFetching } from '@/composables/useFlawsFetching';

import { mountWithConfig } from '@/__tests__/helpers';
import { useSearchStore } from '@/stores/SearchStore';

import IndexView from '../IndexView.vue';

vi.mock('@/composables/useFlawsFetching', () => ({
  useFlawsFetching: vi.fn().mockReturnValue({
    issues: [],
    loadFlaws: vi.fn(),
    isLoading: false,
    isFinalPageFetched: false,
    total: 0,
  }),
}));

vi.mock('@/stores/SearchStore', () => ({
  useSearchStore: vi.fn().mockReturnValue({
    queryFilter: 'django query',
    searchFilters: { affects__ps_component: 'test' },
  }),
}));

describe('indexView', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should render', () => {
    const wrapper = mountWithConfig(IndexView);

    expect(wrapper.findComponent(IssueQueue).exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call loadFlaws on mount with default filter', async () => {
    const { loadFlaws } = useFlawsFetching();
    mountWithConfig(IndexView);

    await flushPromises();

    expect(loadFlaws).toHaveBeenCalledTimes(1);
    expect(loadFlaws).toHaveBeenCalledWith(expect.objectContaining({
      _value: {
        order: '-created_dt',
        affects__ps_component: 'test',
        query: 'django query',
      },
    }));
  });

  it('should call loadFlaws on mount without filters', async () => {
    vi.mocked(useSearchStore, { partial: true }).mockReturnValue({
      queryFilter: '',
      searchFilters: {},
    });
    const { loadFlaws } = useFlawsFetching();
    const wrapper = mountWithConfig(IndexView);

    await wrapper.find('div.osim-incident-filter').findComponent(LabelCheckbox).trigger('click');
    await flushPromises();

    expect(loadFlaws).toHaveBeenCalledTimes(1);
    expect(loadFlaws).toHaveBeenCalledWith(expect.objectContaining({
      _value: {
        order: '-created_dt',
      },
    }));
  });

  it('should call loadFlaws when filters are changed', async () => {
    const { loadFlaws } = useFlawsFetching();
    const wrapper: VueWrapper<Partial<{ setTableFilters: (filters: unknown) => void }>> = mountWithConfig(IndexView);
    await flushPromises();

    wrapper.vm.setTableFilters!(ref({
      order: 'updated_dt',
    }));
    await flushPromises();

    expect(loadFlaws).toHaveBeenCalledTimes(2);
    expect(loadFlaws).toHaveBeenCalledWith(expect.objectContaining({
      _value: {
        order: 'updated_dt',
      },
    }));
  });
});
