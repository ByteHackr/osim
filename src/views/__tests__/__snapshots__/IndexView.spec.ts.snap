// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`indexView > should render 1`] = `
"<main class="mt-3">
  <div data-v-5f4ac70c="" class="osim-content container-fluid osim-issue-queue">
    <div data-v-5f4ac70c="" class="osim-incident-filter">
      <div data-v-5fd32c51="" data-v-5f4ac70c="" class="osim-input ps-3 d-inline-block d-inline-block"><label data-v-5fd32c51="" class="form-check h-100 d-inline-block ps-4 p-1"><input data-v-5fd32c51="" class="d-inline-block form-check-input" type="checkbox"><span data-v-5fd32c51="" class="form-check-label">My Issues</span></label></div>
      <div data-v-5fd32c51="" data-v-5f4ac70c="" class="osim-input ps-3 d-inline-block d-inline-block"><label data-v-5fd32c51="" class="form-check h-100 d-inline-block ps-4 p-1"><input data-v-5fd32c51="" class="d-inline-block form-check-input" type="checkbox"><span data-v-5fd32c51="" class="form-check-label">Open Issues</span></label></div>
      <div data-v-5fd32c51="" data-v-5f4ac70c="" class="osim-input ps-3 d-inline-block d-inline-block"><label data-v-5fd32c51="" class="form-check h-100 d-inline-block ps-4 p-1"><input data-v-5fd32c51="" class="d-inline-block form-check-input" type="checkbox" checked=""><span data-v-5fd32c51="" class="form-check-label">Default Filter</span></label></div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-5f4ac70c="" class="d-flex gap-2">
      <details data-v-5f4ac70c="" class="osim-default-filter">
        <summary data-v-5f4ac70c="">Default Filters <button data-v-5f4ac70c="" class="btn btn-sm btn-primary lh-0 py-0"> clear </button></summary>
        <div data-v-5f4ac70c="" class="my-2"><span data-v-5f4ac70c="" class="badge bg-secondary me-1">Query : django query</span><span data-v-5f4ac70c="" class="badge bg-secondary me-1">Affected Component : test</span></div>
      </details>
    </div>
    <div data-v-5f4ac70c="" class="osim-incident-list">
      <table data-v-5f4ac70c="" class="table align-middle">
        <thead data-v-5f4ac70c="" class="sticky-top">
          <!-- <thead class=""> -->
          <tr data-v-5f4ac70c="">
            <th data-v-5f4ac70c="">ID <i data-v-5f4ac70c="" class="opacity-0 bi-caret-down-fill bi"></i></th>
            <th data-v-5f4ac70c="">Impact <i data-v-5f4ac70c="" class="opacity-0 bi-caret-down-fill bi"></i></th>
            <th data-v-5f4ac70c="">Created <i data-v-5f4ac70c="" class="bi-caret-down-fill bi"></i></th>
            <th data-v-5f4ac70c="">Title <i data-v-5f4ac70c="" class="opacity-0 bi-caret-down-fill bi"></i></th>
            <th data-v-5f4ac70c="">State <i data-v-5f4ac70c="" class="opacity-0 bi-caret-down-fill bi"></i></th>
            <th data-v-5f4ac70c="">Owner <i data-v-5f4ac70c="" class="opacity-0 bi-caret-down-fill bi"></i></th>
          </tr>
        </thead>
        <tbody data-v-5f4ac70c="" class="table-group-divider"></tbody>
      </table>
      <!--v-if-->
      <p data-v-5f4ac70c="">No results.</p>
      <!--v-if-->
    </div>
  </div>
</main>"
`;
